src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -76,12 +128,37 @@ public static SourceLocation Advance(SourceLocation left, string text)
             return tracker.CurrentLocation;
         }
 
-        public static SourceLocation Add(SourceLocation left, SourceLocation right)
+        private static SourceLocation CreateUndefined()
         {
+            var sl = new SourceLocation();
+            sl.AbsoluteIndex = -1;
+            sl.LineIndex = -1;
+            sl.CharacterIndex = -1;
+            return sl;
+        }
+
+        /// <summary>
+        /// Adds two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the sum of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
+        public static SourceLocation operator +(SourceLocation left, SourceLocation right)
```

Comment:
`[NotNull]` for both parameters; generally `public` constructors and methods in this PR need a sweep